temporalio/lib/temporalio/worker/illegal_workflow_call_validator.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+module Temporalio
+  class Worker
+    # Custom validator for validating illegal workflow calls.
+    class IllegalWorkflowCallValidator
+      CallInfo = Data.define(
+        :class_name,
+        :method_name,
+        :trace_point
+      )
+
+      # Call info passed to the validation block for each validation.
+      #
+      # @!attribute class_name
+      #   @return [String] Class name the method is on.
+      # @!attribute method_name
+      #   @return [String] Method name being called.
+      # @!attribute trace_point
+      #   @return [TracePoint] TracePoint instance for the call.
+      class CallInfo; end # rubocop:disable Lint/EmptyClass
+
+      # @return [Array<IllegalWorkflowCallValidator>] Set of advanced validators for Time calls.
+      def self.default_time_validators
+        @default_time_validators ||= [
+          # Do not consider initialize as invalid if year is present and not "true"
+          IllegalWorkflowCallValidator.new(method_name: :initialize) do |info|
+            year_val = info.trace_point.binding&.local_variable_get(:year)
+            raise 'can only use if passing string or explicit time info' unless year_val && year_val != true
+          end,
+          IllegalWorkflowCallValidator.new(method_name: :now) do
+            # When the xmlschema (aliased as iso8601) call is made, zone_offset is called which has a default parameter
+            # of Time.now.year. We want to prevent failing in that specific case. It is expensive to access the caller
+            # stack, but this is only done in the rare case they are calling this safely.
+            next if caller_locations&.any? { |loc| loc.label == 'Time.zone_offset' }
```

Comment:
Do we? I mean, it's technically possible that a WFT isn't picked up across a year boundary. Not even that hard to have it happen if it's happening right around NYE UTC time.

IE: WFT timestamp is just pre-midnight, then replay happens after midnight and we end up assuming it's that time except a year forward. 

Or, is what's actually happening here that the year has to be provided and this call is still made anyway? Not immediately obvious to me.

Reply:
> Do we?

IMO yes. Here's what's happening...

A user reported they can't even parse a time in a workflow (totally deterministic) because of this limitation. The line of code at https://github.com/ruby/ruby/blob/5124f9ac7513eb590c37717337c430cb93caa151/lib/time.rb#L640 is triggering it. During ISO-8601 parse the Ruby code needs zone offset, but obtaining zone offset (https://github.com/ruby/ruby/blob/5124f9ac7513eb590c37717337c430cb93caa151/lib/time.rb#L82) asks for current year it doesn't even really use for these cases.

Reply:
Ah, so, to clarify we're not _actually_ getting the year from that, the call is just made anyway, even though we require the year to explicitly be set or set it ourselves?

Reply:
Exactly. It's a default parameter evaluated at invocation that is not even used in this case.